File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
class Solution {
public:
    int mx = INT_MIN;
    
    int help(TreeNode* root) {
        
        if(!root) return 0;
        
        // take max values from left and right subtrees
        int left = max(0, help(root->left));
        int right = max(0, help(root->right));
        
        // we already have the max values from left and right subtrees
        mx = max(mx, left + root->val + right);
        
        // now return the best path to the parent 
        return max(left, right) + root->val;
    }
    
    int maxPathSum(TreeNode* root) {
        int x = help(root);
        return mx;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

var maxPathSum = function(root) {
    
    let mx = -Infinity;

    let help = (root) => {
        
        if(!root) return 0;

        let left =  Math.max(0, help(root.left))
        let right =  Math.max(0,help(root.right))

        mx = Math.max(mx, (left + right + root.val));
 
        return Math.max(left, right) + root.val;
    }

    help(root)

    return mx;

};